rfc/src/rfcs/0004-loop-contract-synthesis.md

```diff
@@ -0,0 +1,145 @@
+
+The loop-contract synthesizer will synthesize loop contracts for loops without unwinding numbers specified. With the synthesized loop contracts, CBMC abstracts those loops as non-loop structures that can cover arbitrary iterations of the loops. As a result, with the synthesizer proposed in this RFC, Kani can verify programs without unwind loops, and provide a stronger guarantee on correctness. For example, the number of executed iterations of the loop in the following harness is dynamically bounded by an unbounded variable `y`; hence no unwinding number can guarantee the correctness of the loop.
+```rust
+#[kani::proof]
+fn main() {
+    let mut y: i32 = kani::any_where(|i| *i > 0, "y is a positive integer");
+
+    while y > 0 {
+        y = y - 1;
+    }
+    assert!(y == 0);
+}
+```
+With the loop-contract synthesizer, Kani can synthesize the loop invariant `y >= 0`, with which it can prove the post-condition `y == 0` without unwinding the loop.
+
+
+Also, these contracts, when provided, would also significantly improve Kani’s verification time since all loops would be unrolled only to a single iteration, as opposed to a number of iterations which is Kani’s current behavior. For example, although the following loop is bounded and we can easily find out the unwinding number `50000` for it, unwinding the loop 50000 times will lead to a large query, and hence slow down Kani. With loop contract, we only verify the single abstract iteration of the loop, which leads to a smaller query.
+```rust
+#[kani::proof]
+#[kani::unwind(50000)]
+fn main() {
+    let mut y: i32 = 50000;
+
+    while y > 0 {
+        y = y - 1;
+    }
+    assert!(y == 0);
+}
+```
+
+
+## User Experience
+
+Once this RFC has been stabilized, users will be able to synthesize loop contracts for loops and verify programs with the synthesized loop contracts with a new command line flag `--synthesize-loop-contracts`.
+
+### How the Synthesizer Works with Unwinding Numbers
+For each loop, users will be able to choose to unwind it for a certain number of times by specifying the unwinding numbers explicitly, or to synthesize loop contracts for it when no unwinding number specified. 
+
+We introduce a new way for users to specify unwinding numbers for a single loops rather than just for the whole functions: specifying loop numbers using labels. E.g., in the following program, both loops in the function `foo` will be unwound `5` times, while only the loop `inner` in the function `bar` will be unwound `2` times.
+```rust
+#[kani::unwind(5)]
+fn foo() {
+    let mut y: i32 = 50000;
+    let mut x: i32 = 2;
+    while y > 0 {
+        while x > 0 {
+            y = y - 1;
+            x = x - 1;
+        }
+        x = 2;
+    }
+    assert!(y == 0 || y == -1);
+}
+
+#[kani::unwind(inner:2)]
+fn bar() {
+    let mut y: i32 = 50000;
+    let mut x: i32 = 2;
+    `outer`: while y > 0 {
```

Comment:
that's a neat idea... can you please expand how this works when the user also provides the command line `--unwind` option? Also please make it clear what happens if the user uses a loop specific bound and a generic one for the harness.